### PR TITLE
Generate service package data for resource self-registration

### PR DESCRIPTION
**Pull request by ewbankkit**
### Description

A step forward for resource (and data source) self-registration.

* In each service package directory generate `service_package_data_gen.go` which contains an exported implementation of `intf.ServicePackageData` and functions to allow Plugin Framework and Plugin SDK resources and data sources to self-register. Remove the explicit calls to generate these from a small number of services' `generate.go`s.
* Generate `internal/provider/service_packages_gen.go` containing a list of every service package's  exported `ServicePackageData`.
* In `schema.Provider`'s constructor function, use the list of `ServicePackageData`s to augment the static data source and resource maps for the provider.
* If the self-registered Plugin SDK resources (and data sources) implement any `WithoutTimeout` CRUDI handlers, wrap those handlers.
* In the wrapped handlers (and the equivalent for wrapped Plugin Framework resources and data sources), call `AWSClient.InitContext` to create a standard [`context.Context`](https://pkg.go.dev/context#Context) (which currently does nothing extra).

#### Questions

* Is the form of the `service_package_data_gen.go`s OK? Should the `register...` functions be free functions or methods on `servicePackageData`?
* Does generating a `service_package_data_gen.go` for each service package in a single generator (`internal/generate/servicepackagedata/main.go`) make sense? I did it like this so that the same generator can also generate the single list of all `ServicePackageData`s.
* For Plugin SDK resources does changing `func ResourceXYZ() *schema.Resource` to `func resourceXYZ() *schema.Resource` and adding a `var ResourceXYZ = resourceXYZ` to `exports_test.go` (so that it's only externally visible in the test context) make sense?
* Is the CRUDI handler wrapping and centralized `context.Context` creation OK?

#### TODO

* Graduate `internal/service/experimental/intf`
* Update documentation (both internal design and external)

### Relations

Relates https://github.com/hashicorp/terraform-provider-aws/pull/26101.
Relates https://github.com/hashicorp/terraform-provider-aws/pull/27021.
Relates https://github.com/hashicorp/terraform-provider-aws/pull/27221.
Relates https://github.com/hashicorp/terraform-provider-aws/issues/26626.

### Output from Acceptance Testing

```console
% make testacc TESTARGS='-run=TestAccLogsGroup' PKG=logs ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/logs/... -v -count 1 -parallel 2  -run=TestAccLogsGroup -timeout 180m
=== RUN   TestAccLogsGroupDataSource_basic
=== PAUSE TestAccLogsGroupDataSource_basic
=== RUN   TestAccLogsGroup_basic
=== PAUSE TestAccLogsGroup_basic
=== RUN   TestAccLogsGroup_nameGenerate
=== PAUSE TestAccLogsGroup_nameGenerate
=== RUN   TestAccLogsGroup_namePrefix
=== PAUSE TestAccLogsGroup_namePrefix
=== RUN   TestAccLogsGroup_disappears
=== PAUSE TestAccLogsGroup_disappears
=== RUN   TestAccLogsGroup_tags
=== PAUSE TestAccLogsGroup_tags
=== RUN   TestAccLogsGroup_kmsKey
=== PAUSE TestAccLogsGroup_kmsKey
=== RUN   TestAccLogsGroup_retentionPolicy
=== PAUSE TestAccLogsGroup_retentionPolicy
=== RUN   TestAccLogsGroup_multiple
=== PAUSE TestAccLogsGroup_multiple
=== RUN   TestAccLogsGroup_skipDestroy
=== PAUSE TestAccLogsGroup_skipDestroy
=== RUN   TestAccLogsGroupsDataSource_basic
=== PAUSE TestAccLogsGroupsDataSource_basic
=== RUN   TestAccLogsGroupsDataSource_noPrefix
=== PAUSE TestAccLogsGroupsDataSource_noPrefix
=== CONT  TestAccLogsGroupDataSource_basic
=== CONT  TestAccLogsGroup_kmsKey
--- PASS: TestAccLogsGroupDataSource_basic (14.27s)
=== CONT  TestAccLogsGroup_namePrefix
--- PASS: TestAccLogsGroup_namePrefix (15.44s)
=== CONT  TestAccLogsGroup_tags
--- PASS: TestAccLogsGroup_kmsKey (45.58s)
=== CONT  TestAccLogsGroup_disappears
--- PASS: TestAccLogsGroup_disappears (11.03s)
=== CONT  TestAccLogsGroup_nameGenerate
--- PASS: TestAccLogsGroup_tags (36.66s)
=== CONT  TestAccLogsGroup_skipDestroy
--- PASS: TestAccLogsGroup_nameGenerate (16.26s)
=== CONT  TestAccLogsGroupsDataSource_noPrefix
--- PASS: TestAccLogsGroup_skipDestroy (13.87s)
=== CONT  TestAccLogsGroupsDataSource_basic
--- PASS: TestAccLogsGroupsDataSource_basic (14.02s)
=== CONT  TestAccLogsGroup_multiple
--- PASS: TestAccLogsGroupsDataSource_noPrefix (24.57s)
=== CONT  TestAccLogsGroup_retentionPolicy
--- PASS: TestAccLogsGroup_multiple (13.56s)
=== CONT  TestAccLogsGroup_basic
--- PASS: TestAccLogsGroup_basic (15.82s)
--- PASS: TestAccLogsGroup_retentionPolicy (26.28s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/logs	128.762s
% make testacc TESTARGS='-run=TestAccSimpleDBDomain_' PKG=simpledb ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/simpledb/... -v -count 1 -parallel 2  -run=TestAccSimpleDBDomain_ -timeout 180m
=== RUN   TestAccSimpleDBDomain_basic
=== PAUSE TestAccSimpleDBDomain_basic
=== RUN   TestAccSimpleDBDomain_disappears
=== PAUSE TestAccSimpleDBDomain_disappears
=== RUN   TestAccSimpleDBDomain_MigrateFromPluginSDK
=== PAUSE TestAccSimpleDBDomain_MigrateFromPluginSDK
=== CONT  TestAccSimpleDBDomain_basic
=== CONT  TestAccSimpleDBDomain_MigrateFromPluginSDK
--- PASS: TestAccSimpleDBDomain_basic (20.25s)
=== CONT  TestAccSimpleDBDomain_disappears
--- PASS: TestAccSimpleDBDomain_disappears (11.17s)
--- PASS: TestAccSimpleDBDomain_MigrateFromPluginSDK (57.11s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/simpledb	62.011s
```

### CODE REVIEW

File: internal/provider/provider.go [hunk 1]
```diff
@@ -2248,24 +2240,71 @@ func New(_ context.Context) (*schema.Provider, error) {
...
+
+			if v := ds.ReadWithoutTimeout; v != nil {
+				ds.ReadWithoutTimeout = wrappedReadContextFunc(v)
+			}
+
+			provider.DataSourcesMap[typeName] = ds
+		}
+	}
+
+	for _, sp := range servicePackages {
```
Comment (jar-b) [inline internal/provider/provider.go:2265]: Nit - Could this be combined with the range above (unless we intend to add multierror checking between)?

**Review by jar-b — APPROVED** [review]
LGTM 💯

**Comment by YakDriver** [comment]
>* Is the form of the `service_package_data_gen.go`s OK? Should the `register...` functions be free functions or methods on `servicePackageData`?

Would we lose anything by just calling it "service package," dropping the data? `service_package_gen.go` and `servicePackages`?.

Also, I like `register...` being "methods" of `servicePackages` or `servicePackageData`. I'm not sure if there are negative implications down the road but seems like it should be fine.

>* Does generating a `service_package_data_gen.go` for each service package in a single generator (`internal/generate/servicepackagedata/main.go`) make sense? I did it like this so that the same generator can also generate the single list of all `ServicePackageData`s.

A generator of generators is pretty awesome!

>* For Plugin SDK resources does changing `func ResourceXYZ() *schema.Resource` to `func resourceXYZ() *schema.Resource` and adding a `var ResourceXYZ = resourceXYZ` to `exports_test.go` (so that it's only externally visible in the test context) make sense?

I'm not sure about making sense. It seems fine to do. What are we accomplishing with _not_ exporting the resource? Is there some occasion where people might use the resource in a way that they shouldn't? Or is this about simplifying heap during compile?

>* Is the CRUDI handler wrapping and centralized `context.Context` creation OK?

I'll look at this more but don't have any comments.

**Comment by ewbankkit** [comment]
@YakDriver

> I'm not sure about making sense. It seems fine to do. What are we accomplishing with not exporting the resource? Is there some occasion where people might use the resource in a way that they shouldn't? Or is this about simplifying heap during compile?

Yes, the idea here is to help the compiler, reduce the public API surface of the packages and help `golangci-lint` find unused code which had been exported from the package previously.

**Comment by ewbankkit** [comment]
> Would we lose anything by just calling it "service package," dropping the data? `service_package_gen.go` and `servicePackages`?

Make sense.

**Comment by ewbankkit** [comment]
##### TODO

For SDK resources/data sources, allow multiple names for the same resource type.

**Comment by ewbankkit** [comment]
```console
% go test -bench=BenchmarkProtoV5ProviderServerFactory -benchtime 1x -benchmem -run=B -v ./internal/provider
goos: darwin
goarch: amd64
pkg: github.com/hashicorp/terraform-provider-aws/internal/provider
cpu: Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz
BenchmarkProtoV5ProviderServerFactory
    factory_test.go:19: 1125 resources, 425 data sources
BenchmarkProtoV5ProviderServerFactory-8   	       1	 134322926 ns/op	78261224 B/op	 1077607 allocs/op
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/provider	4.803s
```

**Comment by ewbankkit** [comment]
```console

**Pull request merged**